engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java

```diff
@@ -130,7 +122,15 @@ private AbstractHSQuery(ExtendedSearchIntegrator extendedIntegrator, IndexedType
 				throw LOG.targetedEntityTypesNotIndexed( StringHelper.join( targetedEntities, "," ) );
 			}
 		}
-		this.customTypeMetadata = customTypeMetadata == null ? Collections.emptyList() : new ArrayList<>( customTypeMetadata );
+		if ( customTypeMetadata == null ) {
+			this.customTypeMetadata = IndexedTypeMaps.empty();
+		}
+		else {
+			this.customTypeMetadata = IndexedTypeMaps.hashMap();
+			for ( Map.Entry<IndexedTypeIdentifier, CustomTypeMetadata> entry : customTypeMetadata.entrySet() ) {
```

Comment:
looks like you're making a defensive copy here. Why is that?

The reason the modifiers on the `IndexedTypeMap` are deprecated is to eventually make the contract explicitly immutable - granted this is not done yet so we still need to be careful but I think we should work into that direction.